bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/items/ItemUIRegistryImpl.java

```diff
@@ -455,22 +455,76 @@ private Switch createPlayerButtons() {
                 label = label.trim();
-                label = label.substring(0, label.indexOf("[") + 1) + formatPattern + "]";
+                int index = label.indexOf("[");
+                if (index >= 0) {
+                    label = label.substring(0, label.indexOf("[") + 1) + formatPattern + "]";
+                }
             }
         }
 
         return transform(label, considerTransform, labelMappedOption);
     }
 
     private QuantityType<?> convertStateToWidgetUnit(QuantityType<?> quantityState, Widget w) {
-        Unit<?> widgetUnit = UnitUtils.parseUnit(getFormatPattern(w.getLabel()));
+        Unit<?> widgetUnit = UnitUtils.parseUnit(getFormatPattern(w));
         if (widgetUnit != null && !widgetUnit.equals(quantityState.getUnit())) {
             return Objects.requireNonNullElse(quantityState.toInvertibleUnit(widgetUnit), quantityState);
         }
 
         return quantityState;
     }
 
+    @Override
+    public @Nullable String getFormatPattern(Widget w) {
+        String label = getLabelFromWidget(w);
+        String pattern = getFormatPattern(label);
+        String itemName = w.getItem();
+        try {
+            Item item = null;
+            if (itemName != null && !itemName.isBlank()) {
+                item = getItem(itemName);
+            }
+            if (item != null && pattern == null) {
+                StateDescription stateDescription = item.getStateDescription();
+                if (stateDescription != null) {
+                    pattern = stateDescription.getPattern();
+                }
+            }
+
+            if (pattern == null) {
+                return null;
+            }
+
+            // remove last part of pattern, after unit, if it exists, as this is not valid and creates problems with
+            // updates
+            if (item instanceof NumberItem numberItem && numberItem.getDimension() != null) {
+                Matcher m = FORMAT_PATTERN.matcher(pattern);
+                int matcherEnd = 0;
+                while (m.find() && m.group(1) == null) {
+                    matcherEnd = m.end();
+                }
+                String unit = pattern.substring(matcherEnd).trim();
+                String postfix = "";
+                int unitEnd = unit.indexOf(" ");
+                if (unitEnd > -1) {
+                    postfix = unit.substring(unitEnd + 1).trim();
+                    unit = unit.substring(0, unitEnd);
+                }
+                if (!postfix.isBlank()) {
+                    logger.warn(
+                            "Item '{}' with unit, nothing allowed after unit in label pattern '{}', dropping postfix",
+                            itemName, pattern);
+                }
+                pattern = pattern.substring(0, matcherEnd) + (!unit.isBlank() ? " " + unit : "");
+            }
+
+        } catch (ItemNotFoundException e) {
+            logger.error("Cannot retrieve item '{}' for widget {}", itemName, w.eClass().getInstanceTypeName());
```

Comment:
Looks like a too much high log level. WARN level should be sufficient.

Reply:
This wasn't consistent across the class and I just copied one with error level. I will make it warn everywhere.